packages/core/src/CreateNodeAsTool.ts

```diff
@@ -1,182 +1,244 @@
-		const fullPath = currentPath.join('.');
-
-		if (typeof item === 'string' && item.startsWith("{{ '__PLACEHOLDER")) {
-			// Store placeholder values with their full path
-			results.set(encodeDotNotation(fullPath), item);
-		} else if (Array.isArray(item)) {
-			// Recursively traverse nested arrays
-			traverseArray(item, currentPath, results);
-		} else if (typeof item === 'object' && item !== null) {
-			// Recursively traverse nested objects
-			traverseObject(item as NestedObject, currentPath, results);
-		}
-	});
+function parseDefaultValue(value: string | undefined): string | number | boolean | undefined {
+	if (value === undefined) return undefined;
+	if (value === 'true') return true;
+	if (value === 'false') return false;
+	if (!isNaN(Number(value))) return Number(value);
+	return value;
 }
 
 /**
- * Builds a nested object structure from matching keys and their values.
- * @param {string} baseKey - The base key to start building from.
- * @param {string[]} matchingKeys - Array of matching keys.
- * @param {Record<string, string>} values - Object containing values for the keys.
- * @returns {Record<string, unknown>} The built nested object structure.
+ * Parses the arguments of a single fromAI function call.
+ *
+ * This function takes a string representation of function arguments and parses them into
+ * a structured format. It handles complex nested structures, respecting quotes and parentheses.
+ * The function is designed to work with arguments that may contain nested objects, arrays,
+ * or quoted strings, ensuring that these are correctly parsed as single entities.
+ *
+ * @param argsString - The string containing the function arguments, potentially including
+ *                     nested structures and quoted strings
+ * @returns A FromAIArgument object with parsed values, maintaining the structure and types
+ *          of the original arguments
  */
-function buildStructureFromMatches(
-	baseKey: string,
-	matchingKeys: string[],
-	values: Record<string, string>,
-): Record<string, unknown> {
-	const result = {};
+function parseArguments(argsString: string): FromAIArgument {
+	const args: string[] = [];
+	let currentArg = '';
+	let inQuotes = false;
+	let quoteChar = '';
+	let depth = 0;
 
-	for (const matchingKey of matchingKeys) {
-		const decodedKey = decodeDotNotation(matchingKey);
-		// Extract the part of the key after the base key
-		const remainingPath = decodedKey
-			.slice(baseKey.length)
-			.split('.')
-			.filter((k) => k !== '');
-		let current: Record<string, unknown> = result;
+	// This loop iterates through each character in the argsString
+	for (let i = 0; i < argsString.length; i++) {
+		const char = argsString[i];
 
-		// Build the nested structure
-		for (let i = 0; i < remainingPath.length - 1; i++) {
-			if (!(remainingPath[i] in current)) {
-				current[remainingPath[i]] = {};
+		// Handle quotes (single or double)
+		if (char === "'" || char === '"') {
```

Comment:
Does this handle escaped characters properly? I tried it, and got some weird results:

```
argsString: 'database-filter', 'the filter\\'s keyword to apply to the database', string, ['a\\'b', 'b']
result: {
  key: 'database-filter',
  description: "the filter\\\\'s keyword to apply to the database', string, ['a\\\\'b",
  type: "b']",
  defaultValue: undefined
}
```

Reply:
I improved the parsing method and added a few more tests; could you have another look?